{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Graphics including GenomeDiagram {#chapter:graphics}\n",
    "================================\n",
    "\n",
    "The `Bio.Graphics` module depends on the third party Python library\n",
    "[ReportLab](http://www.reportlab.org). Although focused on producing PDF\n",
    "files, ReportLab can also create encapsulated postscript (EPS) and (SVG)\n",
    "files. In addition to these vector based images, provided certain\n",
    "further dependencies such as the [Python Imaging Library\n",
    "(PIL)](http://www.pythonware.com/products/pil/) are installed, ReportLab\n",
    "can also output bitmap images (including JPEG, PNG, GIF, BMP and PICT\n",
    "formats).\n",
    "\n",
    "GenomeDiagram {#sec:genomediagram}\n",
    "-------------\n",
    "\n",
    "### Introduction\n",
    "\n",
    "The `Bio.Graphics.GenomeDiagram` module was added to Biopython 1.50,\n",
    "having previously been available as a separate Python module dependent\n",
    "on Biopython. GenomeDiagram is described in the Bioinformatics journal\n",
    "publication by Pritchard et al. (2006) @pritchard2006, which includes\n",
    "some examples images. There is a PDF copy of the old manual here,\n",
    "<http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf> which has\n",
    "some more examples. As the name might suggest, GenomeDiagram was\n",
    "designed for drawing whole genomes, in particular prokaryotic genomes,\n",
    "either as linear diagrams (optionally broken up into fragments to fit\n",
    "better) or as circular wheel diagrams. Have a look at Figure 2 in Toth\n",
    "*et al.* (2006) @toth2006 for a good example. It proved also well suited\n",
    "to drawing quite detailed figures for smaller genomes such as phage,\n",
    "plasmids or mitochrondia, for example see Figures 1 and 2 in Van der\n",
    "Auwera *et al.* (2009) @vanderauwera2009 (shown with additional manual\n",
    "editing).\n",
    "\n",
    "This module is easiest to use if you have your genome loaded as a\n",
    "`SeqRecord` object containing lots of `SeqFeature` objects - for example\n",
    "as loaded from a GenBank file (see Chapters \\[chapter:SeqRecord\\]\n",
    "and \\[chapter:Bio.SeqIO\\]).\n",
    "\n",
    "### Diagrams, tracks, feature-sets and features\n",
    "\n",
    "GenomeDiagram uses a nested set of objects. At the top level, you have a\n",
    "diagram object representing a sequence (or sequence region) along the\n",
    "horizontal axis (or circle). A diagram can contain one or more tracks,\n",
    "shown stacked vertically (or radially on circular diagrams). These will\n",
    "typically all have the same length and represent the same sequence\n",
    "region. You might use one track to show the gene locations, another to\n",
    "show regulatory regions, and a third track to show the GC percentage.\n",
    "The most commonly used type of track will contain features, bundled\n",
    "together in feature-sets. You might choose to use one feature-set for\n",
    "all your CDS features, and another for tRNA features. This isn’t\n",
    "required - they can all go in the same feature-set, but it makes it\n",
    "easier to update the properties of just selected features (e.g. make all\n",
    "the tRNA features red).\n",
    "\n",
    "There are two main ways to build up a complete diagram. Firstly, the top\n",
    "down approach where you create a diagram object, and then using its\n",
    "methods add track(s), and use the track methods to add feature-set(s),\n",
    "and use their methods to add the features. Secondly, you can create the\n",
    "individual objects separately (in whatever order suits your code), and\n",
    "then combine them.\n",
    "\n",
    "### A top down example {#sec:gd_top_down}\n",
    "\n",
    "We’re going to draw a whole genome from a `SeqRecord` object read in\n",
    "from a GenBank file (see Chapter \\[chapter:Bio.SeqIO\\]). This example\n",
    "uses the pPCP1 plasmid from *Yersinia pestis biovar Microtus*, the file\n",
    "is included with the Biopython unit tests under the GenBank folder, or\n",
    "online\n",
    "[`NC_005816.gb`](http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb)\n",
    "from our website.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We’re using a top down approach, so after loading in our sequence we\n",
    "next create an empty diagram, then add an (empty) track, and to that add\n",
    "an (empty) feature set:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Now the fun part - we take each gene `SeqFeature` object in our\n",
    "`SeqRecord`, and use it to generate a feature on the diagram. We’re\n",
    "going to color them blue, alternating between a dark blue and a light\n",
    "blue.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Now we come to actually making the output file. This happens in two\n",
    "steps, first we call the `draw` method, which creates all the shapes\n",
    "using ReportLab objects. Then we call the `write` method which renders\n",
    "these to the requested file format. Note you can output in multiple file\n",
    "formats:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Also, provided you have the dependencies installed, you can also do\n",
    "bitmaps, for example:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The expected output is shown in Figure \\[fig:plasmid\\_linear\\].\n",
    "\n",
    "![Simple linear diagram for *Yersinia pestis biovar Microtus* plasmid\n",
    "pPCP1.<span\n",
    "data-label=\"fig:plasmid_linear\"></span>](images/plasmid_linear.png)\n",
    "\n",
    "Notice that the `fragments` argument which we set to four controls how\n",
    "many pieces the genome gets broken up into.\n",
    "\n",
    "If you want to do a circular figure, then try this:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The expected output is shown in Figure \\[fig:plasmid\\_circular\\].\n",
    "\n",
    "![Simple circular diagram for *Yersinia pestis biovar Microtus* plasmid\n",
    "pPCP1.<span\n",
    "data-label=\"fig:plasmid_circular\"></span>](images/plasmid_circular.png)\n",
    "\n",
    "These figures are not very exciting, but we’ve only just got started.\n",
    "\n",
    "### A bottom up example\n",
    "\n",
    "Now let’s produce exactly the same figures, but using the bottom up\n",
    "approach. This means we create the different objects directly (and this\n",
    "can be done in almost any order) and then combine them.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You can now call the `draw` and `write` methods as before to produce a\n",
    "linear or circular diagram, using the code at the end of the top-down\n",
    "example above. The figures should be identical.\n",
    "\n",
    "### Features without a SeqFeature {#sec:gd_features_without_seqfeatures}\n",
    "\n",
    "In the above example we used a `SeqRecord`’s `SeqFeature` objects to\n",
    "build our diagram (see also Section \\[sec:seq\\_features\\]). Sometimes\n",
    "you won’t have `SeqFeature` objects, but just the coordinates for a\n",
    "feature you want to draw. You have to create minimal `SeqFeature`\n",
    "object, but this is easy:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "For strand, use `+1` for the forward strand, `-1` for the reverse\n",
    "strand, and `None` for both. Here is a short self contained example:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The top part of the image in the next subsection shows the output\n",
    "\n",
    "The output is shown at the top of Figure \\[fig:gd\\_sigil\\_labels\\]\n",
    "\n",
    "(in the default feature color, pale green).\n",
    "\n",
    "Notice that we have used the `name` argument here to specify the caption\n",
    "text for these features. This is discussed in more detail next.\n",
    "\n",
    "### Feature captions {#sec:gd_feature_captions}\n",
    "\n",
    "Recall we used the following (where `feature` was a `SeqFeature` object)\n",
    "to add a feature to the diagram:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In the example above the `SeqFeature` annotation was used to pick a\n",
    "sensible caption for the features. By default the following possible\n",
    "entries under the `SeqFeature` object’s qualifiers dictionary are used:\n",
    "`gene`, `label`, `name`, `locus_tag`, and `product`. More simply, you\n",
    "can specify a name directly:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In addition to the caption text for each feature’s label, you can also\n",
    "choose the font, position (this defaults to the start of the sigil, you\n",
    "can also choose the middle or at the end) and orientation (for linear\n",
    "diagrams only, where this defaults to rotated by $45$ degrees):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Combining each of these three fragments with the complete example in the\n",
    "previous section should give something like\n",
    "\n",
    "this:\n",
    "\n",
    "\\[fig:gd\\_sigil\\_labels\\]\n",
    "\n",
    "the tracks in Figure \\[fig:gd\\_sigil\\_labels\\].\n",
    "\n",
    "![Simple GenomeDiagram showing label options. The top plot in pale green\n",
    "shows the default label settings (see\n",
    "Section \\[sec:gd\\_features\\_without\\_seqfeatures\\]) while the rest show\n",
    "variations in the label size, position and orientation (see\n",
    "Section \\[sec:gd\\_feature\\_captions\\]). <span\n",
    "data-label=\"fig:gd_sigil_labels\"></span>](images/GD_sigil_labels.png)\n",
    "\n",
    "We’ve not shown it here, but you can also set `label_color` to control\n",
    "the label’s color (used in Section \\[sec:gd\\_nice\\_example\\]).\n",
    "\n",
    "You’ll notice the default font is quite small - this makes sense because\n",
    "you will usually be drawing many (small) features on a page, not just a\n",
    "few large ones as shown here.\n",
    "\n",
    "### Feature sigils {#sec:gd_sigils}\n",
    "\n",
    "The examples above have all just used the default sigil for the feature,\n",
    "a plain box, which was all that was available in the last publicly\n",
    "released standalone version of GenomeDiagram. Arrow sigils were included\n",
    "when GenomeDiagram was added to Biopython 1.50:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Biopython 1.61 added three more sigils,\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "These are shown\n",
    "\n",
    "below.\n",
    "\n",
    "in Figure \\[fig:gd\\_sigils\\].\n",
    "\n",
    "Most sigils fit into a bounding box (as given by the default BOX sigil),\n",
    "either above or below the axis for the forward or reverse strand, or\n",
    "straddling it (double the height) for strand-less features. The BIGARROW\n",
    "sigil is different, always straddling the axis with the direction taken\n",
    "from the feature’s stand.\n",
    "\n",
    "![Simple GenomeDiagram showing different sigils (see\n",
    "Section \\[sec:gd\\_sigils\\])<span\n",
    "data-label=\"fig:gd_sigils\"></span>](images/GD_sigils.png)\n",
    "\n",
    "### Arrow sigils {#sec:gd_arrow_sigils}\n",
    "\n",
    "We introduced the arrow sigils in the previous section. There are two\n",
    "additional options to adjust the shapes of the arrows, firstly the\n",
    "thickness of the arrow shaft, given as a proportion of the height of the\n",
    "bounding box:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The results are shown below:\n",
    "\n",
    "The results are shown in Figure \\[fig:gd\\_sigil\\_arrow\\_shafts\\].\n",
    "\n",
    "![Simple GenomeDiagram showing arrow shaft options (see\n",
    "Section \\[sec:gd\\_arrow\\_sigils\\])<span\n",
    "data-label=\"fig:gd_sigil_arrow_shafts\"></span>](images/GD_sigil_arrow_shafts.png)\n",
    "\n",
    "Secondly, the length of the arrow head - given as a proportion of the\n",
    "height of the bounding box (defaulting to $0.5$, or $50\\%$):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The results are shown below:\n",
    "\n",
    "The results are shown in Figure \\[fig:gd\\_sigil\\_arrow\\_heads\\].\n",
    "\n",
    "![Simple GenomeDiagram showing arrow head options (see\n",
    "Section \\[sec:gd\\_arrow\\_sigils\\])<span\n",
    "data-label=\"fig:gd_sigil_arrow_heads\"></span>](images/GD_sigil_arrow_heads.png)\n",
    "\n",
    "Biopython 1.61 adds a new `BIGARROW` sigil which always stradles the\n",
    "axis, pointing left for the reverse strand or right otherwise:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "All the shaft and arrow head options shown above for the `ARROW` sigil\n",
    "can be used for the `BIGARROW` sigil too.\n",
    "\n",
    "### A nice example {#sec:gd_nice_example}\n",
    "\n",
    "Now let’s return to the pPCP1 plasmid from *Yersinia pestis biovar\n",
    "Microtus*, and the top down approach used in\n",
    "Section \\[sec:gd\\_top\\_down\\], but take advantage of the sigil options\n",
    "we’ve now discussed. This time we’ll use arrows for the genes, and\n",
    "overlay them with strand-less features (as plain boxes) showing the\n",
    "position of some restriction digest sites.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "And the output:\n",
    "\n",
    "The expected output is shown in Figures \\[fig:plasmid\\_linear\\_nice\\]\n",
    "and \\[fig:plasmid\\_circular\\_nice\\].\n",
    "\n",
    "![Linear diagram for *Yersinia pestis biovar Microtus* plasmid pPCP1\n",
    "showing selected restriction digest sites (see\n",
    "Section \\[sec:gd\\_nice\\_example\\]).<span\n",
    "data-label=\"fig:plasmid_linear_nice\"></span>](images/plasmid_linear_nice.png)\n",
    "\n",
    "![Circular diagram for *Yersinia pestis biovar Microtus* plasmid pPCP1\n",
    "showing selected restriction digest sites (see\n",
    "Section \\[sec:gd\\_nice\\_example\\]).<span\n",
    "data-label=\"fig:plasmid_circular_nice\"></span>](images/plasmid_circular_nice.png)\n",
    "\n",
    "### Multiple tracks {#sec:gd_multiple_tracks}\n",
    "\n",
    "All the examples so far have used a single track, but you can have more\n",
    "than one track – for example show the genes on one, and repeat regions\n",
    "on another. In this example we’re going to show three phage genomes side\n",
    "by side to scale, inspired by Figure 6 in Proux <span>*e*t al.</span>\n",
    "(2002) @proux2002. We’ll need the GenBank files for the following three\n",
    "phage:\n",
    "\n",
    "-   `NC_002703` – Lactococcus phage Tuc2009, complete genome\n",
    "    ($38347$ bp)\n",
    "\n",
    "-   `AF323668` – Bacteriophage bIL285, complete genome ($35538$ bp)\n",
    "\n",
    "-   `NC_003212` – *Listeria innocua* Clip11262, complete genome, of\n",
    "    which we are focussing only on integrated prophage 5\n",
    "    (similar length).\n",
    "\n",
    "You can download these using Entrez if you like, see\n",
    "Section \\[sec:efetch\\] for more details. For the third record we’ve\n",
    "worked out where the phage is integrated into the genome, and slice the\n",
    "record to extract it (with the features preserved, see\n",
    "Section \\[sec:SeqRecord-slicing\\]), and must also reverse complement to\n",
    "match the orientation of the first two phage (again preserving the\n",
    "features, see Section \\[sec:SeqRecord-reverse-complement\\]):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The figure we are imitating used different colors for different gene\n",
    "functions. One way to do this is to edit the GenBank file to record\n",
    "color preferences for each feature - something [Sanger’s Artemis\n",
    "editor](http://www.sanger.ac.uk/resources/software/artemis/) does, and\n",
    "which GenomeDiagram should understand. Here however, we’ll just hard\n",
    "code three lists of colors.\n",
    "\n",
    "Note that the annotation in the GenBank files doesn’t exactly match that\n",
    "shown in Proux *et al.*, they have drawn some unannotated genes.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Now to draw them – this time we add three tracks to the diagram, and\n",
    "also notice they are given different start/end values to reflect their\n",
    "different lengths (this requires Biopython 1.59 or later).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The result:\n",
    "\n",
    "The expected output is shown in Figure \\[fig:three\\_track\\_simple\\].\n",
    "\n",
    "![Linear diagram with three tracks for Lactococcus phage Tuc2009\n",
    "(NC\\_002703), bacteriophage bIL285 (AF323668), and prophage 5 from\n",
    "*Listeria innocua* Clip11262 (NC\\_003212) (see\n",
    "Section \\[sec:gd\\_multiple\\_tracks\\]).<span\n",
    "data-label=\"fig:three_track_simple\"></span>](images/three_track_simple.png)\n",
    "\n",
    "I did wonder why in the original manuscript there were no red or orange\n",
    "genes marked in the bottom phage. Another important point is here the\n",
    "phage are shown with different lengths - this is because they are all\n",
    "drawn to the same scale (they *are* different lengths).\n",
    "\n",
    "The key difference from the published figure is they have color-coded\n",
    "links between similar proteins – which is what we will do in the next\n",
    "section.\n",
    "\n",
    "### Cross-Links between tracks {#sec:gd_cross_links}\n",
    "\n",
    "Biopython 1.59 added the ability to draw cross links between tracks -\n",
    "both simple linear diagrams as we will show here, but also linear\n",
    "diagrams split into fragments and circular diagrams.\n",
    "\n",
    "Continuing the example from the previous section inspired by Figure 6\n",
    "from Proux *et al.* 2002 @proux2002, we would need a list of cross links\n",
    "between pairs of genes, along with a score or color to use.\n",
    "Realistically you might extract this from a BLAST file computationally,\n",
    "but here I have manually typed them in.\n",
    "\n",
    "My naming convention continues to refer to the three phage as A, B and\n",
    "C. Here are the links we want to show between A and B, given as a list\n",
    "of tuples (percentage similarity score, gene in A, gene in B).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Likewise for B and C:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "For the first and last phage these identifiers are locus tags, for the\n",
    "middle phage there are no locus tags so I’ve used gene names instead.\n",
    "The following little helper function lets us lookup a feature using\n",
    "either a locus tag or gene name:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We can now turn those list of identifier pairs into SeqFeature pairs,\n",
    "and thus find their location co-ordinates. We can now add all that code\n",
    "and the following snippet to the previous example (just before the\n",
    "`gd_diagram.draw(...)` line – see the finished example script\n",
    "[Proux\\_et\\_al\\_2002\\_Figure\\_6.py](http://biopython.org/SRC/biopython/Doc/examples/Proux_et_al_2002_Figure_6.py)\n",
    "included in the `Doc/examples` folder of the Biopython source code) to\n",
    "add cross links to the figure:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "There are several important pieces to this code. First the\n",
    "`GenomeDiagram` object has a `cross_track_links` attribute which is just\n",
    "a list of `CrossLink` objects. Each `CrossLink` object takes two sets of\n",
    "track-specific co-ordinates (here given as tuples, you can alternatively\n",
    "use a `GenomeDiagram.Feature` object instead). You can optionally supply\n",
    "a colour, border color, and say if this link should be drawn flipped\n",
    "(useful for showing inversions).\n",
    "\n",
    "You can also see how we turn the BLAST percentage identity score into a\n",
    "colour, interpolating between white ($0\\%$) and a dark red ($100\\%$). In\n",
    "this example we don’t have any problems with overlapping cross-links.\n",
    "One way to tackle that is to use transparency in ReportLab, by using\n",
    "colors with their alpha channel set. However, this kind of shaded color\n",
    "scheme combined with overlap transparency would be difficult to\n",
    "interpret.\n",
    "\n",
    "The result:\n",
    "\n",
    "The expected output is shown in Figure \\[fig:three\\_track\\_cl\\].\n",
    "\n",
    "![Linear diagram with three tracks for Lactococcus phage Tuc2009\n",
    "(NC\\_002703), bacteriophage bIL285 (AF323668), and prophage 5 from\n",
    "*Listeria innocua* Clip11262 (NC\\_003212) plus basic cross-links shaded\n",
    "by percentage identity (see Section \\[sec:gd\\_cross\\_links\\]).<span\n",
    "data-label=\"fig:three_track_cl\"></span>](images/three_track_cl.png)\n",
    "\n",
    "There is still a lot more that can be done within Biopython to help\n",
    "improve this figure. First of all, the cross links in this case are\n",
    "between proteins which are drawn in a strand specific manor. It can help\n",
    "to add a background region (a feature using the ‘BOX’ sigil) on the\n",
    "feature track to extend the cross link. Also, we could reduce the\n",
    "vertical height of the feature tracks to allocate more to the links\n",
    "instead – one way to do that is to allocate space for empty tracks.\n",
    "Furthermore, in cases like this where there are no large gene overlaps,\n",
    "we can use the axis-straddling `BIGARROW` sigil, which allows us to\n",
    "further reduce the vertical space needed for the track. These\n",
    "improvements are demonstrated in the example script\n",
    "[Proux\\_et\\_al\\_2002\\_Figure\\_6.py](http://biopython.org/SRC/biopython/Doc/examples/Proux_et_al_2002_Figure_6.py)\n",
    "included in the `Doc/examples` folder of the Biopython source code.\n",
    "\n",
    "The result:\n",
    "\n",
    "The expected output is shown in Figure \\[fig:three\\_track\\_cl2\\].\n",
    "\n",
    "![Linear diagram with three tracks for Lactococcus phage Tuc2009\n",
    "(NC\\_002703), bacteriophage bIL285 (AF323668), and prophage 5 from\n",
    "*Listeria innocua* Clip11262 (NC\\_003212) plus cross-links shaded by\n",
    "percentage identity (see Section \\[sec:gd\\_cross\\_links\\]).<span\n",
    "data-label=\"fig:three_track_cl2\"></span>](images/three_track_cl2a.png)\n",
    "\n",
    "Beyond that, finishing touches you might want to do manually in a vector\n",
    "image editor include fine tuning the placement of gene labels, and\n",
    "adding other custom annotation such as highlighting particular regions.\n",
    "\n",
    "Although not really necessary in this example since none of the\n",
    "cross-links overlap, using a transparent color in ReportLab is a very\n",
    "useful technique for superimposing multiple links. However, in this case\n",
    "a shaded color scheme should be avoided.\n",
    "\n",
    "### Further options\n",
    "\n",
    "You can control the tick marks to show the scale – after all every graph\n",
    "should show its units, and the number of the grey-track labels.\n",
    "\n",
    "Also, we have only used the `FeatureSet` so far. GenomeDiagram also has\n",
    "a `GraphSet` which can be used for show line graphs, bar charts and heat\n",
    "plots (e.g. to show plots of GC% on a track parallel to the features).\n",
    "\n",
    "These options are not covered here yet, so for now we refer you to the\n",
    "[\\\n",
    "](http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf)<span>User\n",
    "Guide (PDF)</span> included with the standalone version of GenomeDiagram\n",
    "(but please read the next section first), and the docstrings.\n",
    "\n",
    "### Converting old code\n",
    "\n",
    "If you have old code written using the standalone version of\n",
    "GenomeDiagram, and you want to switch it over to using the new version\n",
    "included with Biopython then you will have to make a few changes - most\n",
    "importantly to your import statements.\n",
    "\n",
    "Also, the older version of GenomeDiagram used only the UK spellings of\n",
    "color and center (colour and centre). You will need to change to the\n",
    "American spellings, although for several years the Biopython version of\n",
    "GenomeDiagram supported both.\n",
    "\n",
    "For example, if you used to have:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "you could just switch the import statements like this:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "and hopefully that should be enough. In the long term you might want to\n",
    "switch to the new names, but you would have to change more of your code:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "or:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "If you run into difficulties, please ask on the Biopython mailing list\n",
    "for advice. One catch is that we have not included the old module\n",
    "`GenomeDiagram.GDUtilities` yet. This included a number of GC% related\n",
    "functions, which will probably be merged under `Bio.SeqUtils` later on.\n",
    "\n",
    "Chromosomes\n",
    "-----------\n",
    "\n",
    "The `Bio.Graphics.BasicChromosome` module allows drawing of chromosomes.\n",
    "There is an example in Jupe *et al.* (2012) @jupe2012 (open access)\n",
    "using colors to highlight different gene families.\n",
    "\n",
    "### Simple Chromosomes\n",
    "\n",
    "Here is a very simple example - for which we’ll use *Arabidopsis\n",
    "thaliana*.\n",
    "\n",
    "![Simple chromosome diagram for *Arabidopsis thaliana*.<span\n",
    "data-label=\"fig:simplechromosome\"></span>](images/simple_chrom.pdf)\n",
    "\n",
    "![Chromosome diagram for *Arabidopsis thaliana* showing tRNA genes.<span\n",
    "data-label=\"fig:trnachromosome\"></span>](images/tRNA_chrom.pdf)\n",
    "\n",
    "You can skip this bit, but first I downloaded the five sequenced\n",
    "chromosomes from the NCBI’s FTP site\n",
    "<ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana> and then\n",
    "parsed them with `Bio.SeqIO` to find out their lengths. You could use\n",
    "the GenBank files for this, but it is faster to use the FASTA files for\n",
    "the whole chromosomes:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "This gave the lengths of the five chromosomes, which we’ll now use in\n",
    "the following short demonstration of the `BasicChromosome` module:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "This should create a very simple PDF file, shown\n",
    "\n",
    "here:\n",
    "\n",
    "in Figure \\[fig:simplechromosome\\].\n",
    "\n",
    "This example is deliberately short and sweet. The next example shows the\n",
    "location of features of interest.\n",
    "\n",
    "### Annotated Chromosomes\n",
    "\n",
    "Continuing from the previous example, let’s also show the tRNA genes.\n",
    "We’ll get their locations by parsing the GenBank files for the five\n",
    "*Arabidopsis thaliana* chromosomes. You’ll need to download these files\n",
    "from the NCBI FTP site\n",
    "<ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana>, and preserve\n",
    "the subdirectory names or edit the paths below:\n",
    "\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 0
}