{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bio.PopGen: Population genetics\n",
    "===============================\n",
    "\n",
    "Bio.PopGen is a Biopython module supporting population genetics,\n",
    "available in Biopython 1.44 onwards.\n",
    "\n",
    "The medium term objective for the module is to support widely used data\n",
    "formats, applications and databases. This module is currently under\n",
    "intense development and support for new features should appear at a\n",
    "rather fast pace. Unfortunately this might also entail some instability\n",
    "on the API, especially if you are using a development version. APIs that\n",
    "are made available on our official public releases should be much more\n",
    "stable.\n",
    "\n",
    "GenePop\n",
    "-------\n",
    "\n",
    "GenePop (<http://genepop.curtin.edu.au/>) is a popular population\n",
    "genetics software package supporting Hardy-Weinberg tests, linkage\n",
    "desiquilibrium, population diferentiation, basic statistics, $F_{st}$\n",
    "and migration estimates, among others. GenePop does not supply sequence\n",
    "based statistics as it doesn’t handle sequence data. The GenePop file\n",
    "format is supported by a wide range of other population genetic software\n",
    "applications, thus making it a relevant format in the population\n",
    "genetics field.\n",
    "\n",
    "Bio.PopGen provides a parser and generator of GenePop file format.\n",
    "Utilities to manipulate the content of a record are also provided. Here\n",
    "is an example on how to read a GenePop file (you can find example\n",
    "GenePop data files in the Test/PopGen directory of Biopython):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "This will read a file called example.gen and parse it. If you do print\n",
    "rec, the record will be output again, in GenePop format.\n",
    "\n",
    "The most important information in rec will be the loci names and\n",
    "population information (but there is more – use help(GenePop.Record) to\n",
    "check the API documentation). Loci names can be found on rec.loci\\_list.\n",
    "Population information can be found on rec.populations. Populations is a\n",
    "list with one element per population. Each element is itself a list of\n",
    "individuals, each individual is a pair composed by individual name and a\n",
    "list of alleles (2 per marker), here is an example for rec.populations:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "So we have two populations, the first with two individuals, the second\n",
    "with only one. The first individual of the first population is called\n",
    "Ind1, allelic information for each of the 3 loci follows. Please note\n",
    "that for any locus, information might be missing (see as an example,\n",
    "Ind2 above).\n",
    "\n",
    "A few utility functions to manipulate GenePop records are made\n",
    "available, here is an example:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "GenePop does not support population names, a limitation which can be\n",
    "cumbersome at times. Functionality to enable population names is\n",
    "currently being planned for Biopython. These extensions won’t break\n",
    "compatibility in any way with the standard format. In the medium term,\n",
    "we would also like to support the GenePop web service.\n",
    "\n",
    "Coalescent simulation\n",
    "---------------------\n",
    "\n",
    "A coalescent simulation is a backward model of population genetics with\n",
    "relation to time. A simulation of ancestry is done until the Most Recent\n",
    "Common Ancestor (MRCA) is found. This ancestry relationship starting on\n",
    "the MRCA and ending on the current generation sample is sometimes called\n",
    "a genealogy. Simple cases assume a population of constant size in time,\n",
    "haploidy, no population structure, and simulate the alleles of a single\n",
    "locus under no selection pressure.\n",
    "\n",
    "Coalescent theory is used in many fields like selection detection,\n",
    "estimation of demographic parameters of real populations or disease gene\n",
    "mapping.\n",
    "\n",
    "The strategy followed in the Biopython implementation of the coalescent\n",
    "was not to create a new, built-in, simulator from scratch but to use an\n",
    "existing one, Fastsimcoal2\n",
    "(<http://cmpg.unibe.ch/software/fastsimcoal2/>). Fastsimcoal2 allows\n",
    "for, among others, population structure, multiple demographic events,\n",
    "simulation of multiple types of loci (SNPs, sequences,\n",
    "STRs/microsatellites and RFLPs) with recombination, diploidy multiple\n",
    "chromosomes or ascertainment bias. Notably Fastsimcoal2 doesn’t support\n",
    "any selection model. We recommend reading Fastsimcoal2’s documentation,\n",
    "available in the link above.\n",
    "\n",
    "The input for Fastsimcoal2 is a file specifying the desired demography\n",
    "and genome, the output is a set of files (typically around 1000) with\n",
    "the simulated genomes of a sample of individuals per subpopulation. This\n",
    "set of files can be used in many ways, like to compute confidence\n",
    "intervals where which certain statistics (e.g., $F_{st}$ or Tajima D)\n",
    "are expected to lie. Real population genetics datasets statistics can\n",
    "then be compared to those confidence intervals.\n",
    "\n",
    "Biopython coalescent code allows to create demographic scenarios and\n",
    "genomes and to run Fastsimcoal2.\n",
    "\n",
    "### Creating scenarios\n",
    "\n",
    "Creating a scenario involves both creating a demography and a chromosome\n",
    "structure. In many cases (e.g. when doing Approximate Bayesian\n",
    "Computations – ABC) it is important to test many parameter variations\n",
    "(e.g. vary the effective population size, $N_e$, between 10, 50, 500 and\n",
    "1000 individuals). The code provided allows for the simulation of\n",
    "scenarios with different demographic parameters very easily.\n",
    "\n",
    "Below we see how we can create scenarios and then how simulate them.\n",
    "\n",
    "#### Demography\n",
    "\n",
    "A few predefined demographies are built-in, all have two shared\n",
    "parameters: sample size (called sample\\_size on the template, see below\n",
    "for its use) per deme and deme size, i.e. subpopulation size\n",
    "(pop\\_size). All demographies are available as templates where all\n",
    "parameters can be varied, each template has a system name. The\n",
    "prefedined demographies/templates are:\n",
    "\n",
    "Single population, constant size\n",
    "\n",
    ":   The standard parameters are enough to specify it. Template\n",
    "    name: simple.\n",
    "\n",
    "Single population, bottleneck\n",
    "\n",
    ":   As seen on figure \\[fig:bottle\\]. The parameters are current\n",
    "    population size (pop\\_size on template ne3 on figure), time of\n",
    "    expansion, given as the generation in the past when it occurred\n",
    "    (expand\\_gen), effective population size during bottleneck (ne2),\n",
    "    time of contraction (contract\\_gen) and original size in the remote\n",
    "    past (ne3). Template name: bottle.\n",
    "\n",
    "Island model\n",
    "\n",
    ":   The typical island model. The total number of demes is specified by\n",
    "    total\\_demes and the migration rate by mig. Template name island.\n",
    "\n",
    "Stepping stone model - 1 dimension\n",
    "\n",
    ":   The stepping stone model in 1 dimension, extremes disconnected. The\n",
    "    total number of demes is total\\_demes, migration rate is mig.\n",
    "    Template name is ssm\\_1d.\n",
    "\n",
    "Stepping stone model - 2 dimensions\n",
    "\n",
    ":   The stepping stone model in 2 dimensions, extremes disconnected. The\n",
    "    parameters are x for the horizontal dimension and y for the vertical\n",
    "    (being the total number of demes x times y), migration rate is mig.\n",
    "    Template name is ssm\\_2d.\n",
    "\n",
    "\\[fig:bottle\\]\n",
    "\n",
    "![A bottleneck<span data-label=\"fig:bottle\"></span>](images/bottle.png)\n",
    "\n",
    "In our first example, we will generate a template for a single\n",
    "population, constant size model with a sample size of 30 and a deme size\n",
    "of 500. The code for this is:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Executing this code snippet will generate a file on the current\n",
    "directory called simple\\_100\\_300.par this file can be given as input to\n",
    "Fastsimcoal2 to simulate the demography (below we will see how Biopython\n",
    "can take care of calling Fastsimcoal2).\n",
    "\n",
    "This code consists of a single function call, let’s discuss it parameter\n",
    "by parameter.\n",
    "\n",
    "The first parameter is the template id (from the list above). We are\n",
    "using the id ’simple’ which is the template for a single population of\n",
    "constant size along time.\n",
    "\n",
    "The second parameter is the chromosome structure. Please ignore it for\n",
    "now, it will be explained in the next section.\n",
    "\n",
    "The third parameter is a list of all required parameters (recall that\n",
    "the simple model only needs sample\\_size and pop\\_size) and possible\n",
    "values (in this case each parameter only has a possible value).\n",
    "\n",
    "Now, let’s consider an example where we want to generate several island\n",
    "models, and we are interested in varying the number of demes: 10, 50 and\n",
    "100 with a migration rate of 1%. Sample size and deme size will be the\n",
    "same as before. Here is the code:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In this case, 3 files will be generated: island\\_100\\_0.01\\_100\\_30.par,\n",
    "island\\_10\\_0.01\\_100\\_30.par and island\\_50\\_0.01\\_100\\_30.par. Notice\n",
    "the rule to make file names: template name, followed by parameter values\n",
    "in reverse order.\n",
    "\n",
    "A few, arguably more esoteric template demographies exist (please check\n",
    "the Bio/PopGen/SimCoal/data directory on Biopython source tree).\n",
    "Furthermore it is possible for the user to create new templates. That\n",
    "functionality will be discussed in a future version of this document.\n",
    "\n",
    "#### Chromosome structure\n",
    "\n",
    "We strongly recommend reading Fastsimcoal2 documentation to understand\n",
    "the full potential available in modeling chromosome structures. In this\n",
    "subsection we only discuss how to implement chromosome structures using\n",
    "the Biopython interface, not the underlying Fastsimcoal2 capabilities.\n",
    "\n",
    "We will start by implementing a single chromosome, with 24 SNPs with a\n",
    "recombination rate immediately on the right of each locus of 0.0005 and\n",
    "a minimum frequency of the minor allele of 0. This will be specified by\n",
    "the following list (to be passed as second parameter to the function\n",
    "generate\\_simcoal\\_from\\_template):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "This is actually the chromosome structure used in the above examples.\n",
    "\n",
    "The chromosome structure is represented by a list of chromosomes, each\n",
    "chromosome (i.e., each element in the list) is composed by a tuple (a\n",
    "pair): the first element is the number of times the chromosome is to be\n",
    "repeated (as there might be interest in repeating the same chromosome\n",
    "many times). The second element is a list of the actual components of\n",
    "the chromosome. Each element is again a pair, the first member is the\n",
    "locus type and the second element the parameters for that locus type.\n",
    "Confused? Before showing more examples let’s review the example above:\n",
    "We have a list with one element (thus one chromosome), the chromosome is\n",
    "a single instance (therefore not to be repeated), it is composed of 24\n",
    "SNPs, with a recombination rate of 0.0005 between each consecutive SNP,\n",
    "the minimum frequency of the minor allele is 0.0 (i.e, it can be absent\n",
    "from a certain population).\n",
    "\n",
    "Let’s see a more complicated example:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We start by having 5 chromosomes with the same structure as above (i.e.,\n",
    "24 SNPs). We then have 2 chromosomes which have a DNA sequence with 10\n",
    "nucleotides, 0.0 recombination rate, 0.0005 mutation rate, and a\n",
    "transition rate of 0.33. Then we have an RFLP with 0.0 recombination\n",
    "rate to the next locus and a 0.0001 mutation rate. Finally we have a\n",
    "microsatellite (or STR), with 0.0 recombination rate to the next locus\n",
    "(note, that as this is a single microsatellite which has no loci\n",
    "following, this recombination rate here is irrelevant), with a mutation\n",
    "rate of 0.001, geometric parameter of 0.0 and a range constraint of 0.0\n",
    "(for information about this parameters please consult the Fastsimcoal2\n",
    "documentation, you can use them to simulate various mutation models,\n",
    "including the typical – for microsatellites – stepwise mutation model\n",
    "among others).\n",
    "\n",
    "### Running Fastsimcoal2\n",
    "\n",
    "We now discuss how to run Fastsimcoal2 from inside Biopython. It is\n",
    "required that the binary for Fastsimcoal2 is called fastsimcoal21 (or\n",
    "fastsimcoal21.exe on Windows based platforms), please note that the\n",
    "typical name when downloading the program is in the format\n",
    "fastsimcoal2\\_x\\_y. As such, when installing Fastsimcoal2 you will need\n",
    "to rename of the downloaded executable so that Biopython can find it.\n",
    "\n",
    "It is possible to run Fastsimcoal2 on files that were not generated\n",
    "using the method above (e.g., writing a parameter file by hand), but we\n",
    "will show an example by creating a model using the framework presented\n",
    "above.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The lines of interest are the last two (plus the new import). Firstly a\n",
    "controller for the application is created. The directory where the\n",
    "binary is located has to be specified.\n",
    "\n",
    "The simulator is then run on the last line: we know, from the rules\n",
    "explained above, that the input file name is simple\\_100\\_30.par for the\n",
    "simulation parameter file created. We then specify that we want to run\n",
    "50 independent simulations, by default Biopython requests a simulation\n",
    "of diploid data, but a third parameter can be added to simulate haploid\n",
    "data (adding as a parameter the string ’0’). Fastsimcoal2 will now run\n",
    "(please note that this can take quite a lot of time) and will create a\n",
    "directory with the simulation results. The results can now be analysed\n",
    "(typically studying the data with Arlequin3). In the future Biopython\n",
    "might support reading the Arlequin3 format and thus allowing for the\n",
    "analysis of Fastsimcoal2 data inside Biopython.\n",
    "\n",
    "Other applications\n",
    "------------------\n",
    "\n",
    "Here we discuss interfaces and utilities to deal with population\n",
    "genetics’ applications which arguably have a smaller user base.\n",
    "\n",
    "### FDist: Detecting selection and molecular adaptation\n",
    "\n",
    "FDist is a selection detection application suite based on computing\n",
    "(i.e. simulating) a “neutral” confidence interval based on $F_{st}$ and\n",
    "heterozygosity. Markers (which can be SNPs, microsatellites, AFLPs among\n",
    "others) which lie outside the “neutral” interval are to be considered as\n",
    "possible candidates for being under selection.\n",
    "\n",
    "FDist is mainly used when the number of markers is considered enough to\n",
    "estimate an average $F_{st}$, but not enough to either have outliers\n",
    "calculated from the dataset directly or, with even more markers for\n",
    "which the relative positions in the genome are known, to use approaches\n",
    "based on, e.g., Extended Haplotype Heterozygosity (EHH).\n",
    "\n",
    "The typical usage pattern for FDist is as follows:\n",
    "\n",
    "1.  Import a dataset from an external format into FDist format.\n",
    "\n",
    "2.  Compute average $F_{st}$. This is done by datacal inside FDist.\n",
    "\n",
    "3.  Simulate “neutral” markers based on the average $F_{st}$ and\n",
    "    expected number of total populations. This is the core operation,\n",
    "    done by fdist inside FDist.\n",
    "\n",
    "4.  Calculate the confidence interval, based on the desired confidence\n",
    "    boundaries (typically 95% or 99%). This is done by cplot and is\n",
    "    mainly used to plot the interval.\n",
    "\n",
    "5.  Assess each marker status against the simulation “neutral”\n",
    "    confidence interval. Done by pv. This is used to detect the outlier\n",
    "    status of each marker against the simulation.\n",
    "\n",
    "We will now discuss each step with illustrating example code (for this\n",
    "example to work FDist binaries have to be on the executable PATH).\n",
    "\n",
    "The FDist data format is application specific and is not used at all by\n",
    "other applications, as such you will probably have to convert your data\n",
    "for use with FDist. Biopython can help you do this. Here is an example\n",
    "converting from GenePop format to FDist format (along with imports that\n",
    "will be needed on examples further below):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "In this code we simply parse a GenePop file and convert it to a FDist\n",
    "record.\n",
    "\n",
    "Printing an FDist record will generate a string that can be directly\n",
    "saved to a file and supplied to FDist. FDist requires the input file to\n",
    "be called infile, therefore we save the record on a file with that name.\n",
    "\n",
    "The most important fields on a FDist record are: num\\_pops, the number\n",
    "of populations; num\\_loci, the number of loci and loci\\_data with the\n",
    "marker data itself. Most probably the details of the record are of no\n",
    "interest to the user, as the record only purpose is to be passed to\n",
    "FDist.\n",
    "\n",
    "The next step is to calculate the average $F_{st}$ of the dataset (along\n",
    "with the sample size):\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "On the first line we create an object to control the call of FDist\n",
    "suite, this object will be used further on in order to call other suite\n",
    "applications.\n",
    "\n",
    "On the second line we call the datacal application which computes the\n",
    "average $F_{st}$ and the sample size. It is worth noting that the\n",
    "$F_{st}$ computed by datacal is a *variation* of Weir and Cockerham’s\n",
    "$\\theta$.\n",
    "\n",
    "We can now call the main fdist application in order to simulate neutral\n",
    "markers.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "npops\n",
    "\n",
    ":   Number of populations existing in nature. This is really a\n",
    "    “guestimate”. Has to be lower than 100.\n",
    "\n",
    "nsamples\n",
    "\n",
    ":   Number of populations sampled, has to be lower than npops.\n",
    "\n",
    "fst\n",
    "\n",
    ":   Average $F_{st}$.\n",
    "\n",
    "sample\\_size\n",
    "\n",
    ":   Average number of individuals sampled on each population.\n",
    "\n",
    "mut\n",
    "\n",
    ":   Mutation model: 0 - Infinite alleles; 1 - Stepwise mutations\n",
    "\n",
    "num\\_sims\n",
    "\n",
    ":   Number of simulations to perform. Typically a number around 40000\n",
    "    will be OK, but if you get a confidence interval that looks sharp\n",
    "    (this can be detected when plotting the confidence interval\n",
    "    computed below) the value can be increased (a suggestion would be\n",
    "    steps of 10000 simulations).\n",
    "\n",
    "The confusion in wording between number of samples and sample size stems\n",
    "from the original application.\n",
    "\n",
    "A file named out.dat will be created with the simulated heterozygosities\n",
    "and $F_{st}$s, it will have as many lines as the number of simulations\n",
    "requested.\n",
    "\n",
    "Note that fdist returns the average $F_{st}$ that it was *capable* of\n",
    "simulating, for more details about this issue please read below the\n",
    "paragraph on approximating the desired average $F_{st}$.\n",
    "\n",
    "The next (optional) step is to calculate the confidence interval:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You can only call cplot after having run fdist.\n",
    "\n",
    "This will calculate the confidence intervals (99% in this case) for a\n",
    "previous fdist run. A list of quadruples is returned. The first element\n",
    "represents the heterozygosity, the second the lower bound of $F_{st}$\n",
    "confidence interval for that heterozygosity, the third the average and\n",
    "the fourth the upper bound. This can be used to trace the confidence\n",
    "interval contour. This list is also written to a file, out.cpl.\n",
    "\n",
    "The main purpose of this step is return a set of points which can be\n",
    "easily used to plot a confidence interval. It can be skipped if the\n",
    "objective is only to assess the status of each marker against the\n",
    "simulation, which is the next step...\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You can only call cplot after having run datacal and fdist.\n",
    "\n",
    "This will use the simulated markers to assess the status of each\n",
    "individual real marker. A list, in the same order than the loci\\_list\n",
    "that is on the FDist record (which is in the same order that the GenePop\n",
    "record) is returned. Each element in the list is a quadruple, the\n",
    "fundamental member of each quadruple is the last element (regarding the\n",
    "other elements, please refer to the pv documentation – for the sake of\n",
    "simplicity we will not discuss them here) which returns the probability\n",
    "of the simulated $F_{st}$ being lower than the marker $F_{st}$. Higher\n",
    "values would indicate a stronger candidate for positive selection, lower\n",
    "values a candidate for balancing selection, and intermediate values a\n",
    "possible neutral marker. What is “higher”, “lower” or “intermediate” is\n",
    "really a subjective issue, but taking a “confidence interval” approach\n",
    "and considering a 95% confidence interval, “higher” would be between\n",
    "0.95 and 1.0, “lower” between 0.0 and 0.05 and “intermediate” between\n",
    "0.05 and 0.95.\n",
    "\n",
    "#### Approximating the desired average $F_{st}$\n",
    "\n",
    "Fdist tries to approximate the desired average $F_{st}$ by doing a\n",
    "coalescent simulation using migration rates based on the formula\n",
    "\n",
    "$$N_{m} = \\frac{1 - F_{st}}{4F_{st}}$$\n",
    "\n",
    "This formula assumes a few premises like an infinite number of\n",
    "populations.\n",
    "\n",
    "In practice, when the number of populations is low, the mutation model\n",
    "is stepwise and the sample size increases, fdist will not be able to\n",
    "simulate an acceptable approximate average $F_{st}$.\n",
    "\n",
    "To address that, a function is provided to iteratively approach the\n",
    "desired value by running several fdists in sequence. This approach is\n",
    "computationally more intensive than running a single fdist run, but\n",
    "yields good results. The following code runs fdist approximating the\n",
    "desired $F_{st}$:\n",
    "\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 0
}