{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick Start – What can you do with Biopython? {#chapter:quick-start}\n",
    "=============================================\n",
    "\n",
    "This section is designed to get you started quickly with Biopython, and\n",
    "to give a general overview of what is available and how to use it. All\n",
    "of the examples in this section assume that you have some general\n",
    "working knowledge of Python, and that you have successfully installed\n",
    "Biopython on your system. If you think you need to brush up on your\n",
    "Python, the main Python web site provides quite a bit of free\n",
    "documentation to get started with (<http://www.python.org/doc/>).\n",
    "\n",
    "Since much biological work on the computer involves connecting with\n",
    "databases on the internet, some of the examples will also require a\n",
    "working internet connection in order to run.\n",
    "\n",
    "Now that that is all out of the way, let’s get into what we can do with\n",
    "Biopython.\n",
    "\n",
    "General overview of what Biopython provides\n",
    "-------------------------------------------\n",
    "\n",
    "As mentioned in the introduction, Biopython is a set of libraries to\n",
    "provide the ability to deal with “things” of interest to biologists\n",
    "working on the computer. In general this means that you will need to\n",
    "have at least some programming experience (in Python, of course!) or at\n",
    "least an interest in learning to program. Biopython’s job is to make\n",
    "your job easier as a programmer by supplying reusable libraries so that\n",
    "you can focus on answering your specific question of interest, instead\n",
    "of focusing on the internals of parsing a particular file format (of\n",
    "course, if you want to help by writing a parser that doesn’t exist and\n",
    "contributing it to Biopython, please go ahead!). So Biopython’s job is\n",
    "to make you happy!\n",
    "\n",
    "One thing to note about Biopython is that it often provides multiple\n",
    "ways of “doing the same thing.” Things have improved in recent releases,\n",
    "but this can still be frustrating as in Python there should ideally be\n",
    "one right way to do something. However, this can also be a real benefit\n",
    "because it gives you lots of flexibility and control over the libraries.\n",
    "The tutorial helps to show you the common or easy ways to do things so\n",
    "that you can just make things work. To learn more about the alternative\n",
    "possibilities, look in the Cookbook (Chapter \\[chapter:cookbook\\], this\n",
    "has some cools tricks and tips), the Advanced section\n",
    "(Chapter \\[chapter:advanced\\]), the built in “docstrings” (via the\n",
    "Python help command, or the [API\n",
    "documentation](http://biopython.org/DIST/docs/api/)) or ultimately the\n",
    "code itself.\n",
    "\n",
    "Working with sequences {#sec:sequences}\n",
    "----------------------\n",
    "\n",
    "Disputably (of course!), the central object in bioinformatics is the\n",
    "sequence. Thus, we’ll start with a quick introduction to the Biopython\n",
    "mechanisms for dealing with sequences, the `Seq` object, which we’ll\n",
    "discuss in more detail in Chapter \\[chapter:Bio.Seq\\].\n",
    "\n",
    "Most of the time when we think about sequences we have in my mind a\n",
    "string of letters like ‘`AGTACACTGGT`’. You can create such `Seq` object\n",
    "with this sequence as follows - the “`>>>`” represents the Python prompt\n",
    "followed by what you would type in:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "my_seq = Seq(\"AGTACACTGGT\")\n",
    "my_seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(my_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_seq.alphabet\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "What we have here is a sequence object with a *generic* alphabet -\n",
    "reflecting the fact we have *not* specified if this is a DNA or protein\n",
    "sequence (okay, a protein with a lot of Alanines, Glycines, Cysteines\n",
    "and Threonines!). We’ll talk more about alphabets in\n",
    "Chapter \\[chapter:Bio.Seq\\].\n",
    "\n",
    "In addition to having an alphabet, the `Seq` object differs from the\n",
    "Python string in the methods it supports. You can’t do this with a plain\n",
    "string:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_seq.complement()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_seq.reverse_complement()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The next most important class is the `SeqRecord` or Sequence Record.\n",
    "This holds a sequence (as a `Seq` object) with additional annotation\n",
    "including an identifier, name and description. The `Bio.SeqIO` module\n",
    "for reading and writing sequence file formats works with `SeqRecord`\n",
    "objects, which will be introduced below and covered in more detail by\n",
    "Chapter \\[chapter:Bio.SeqIO\\].\n",
    "\n",
    "This covers the basic features and uses of the Biopython sequence class.\n",
    "Now that you’ve got some idea of what it is like to interact with the\n",
    "Biopython libraries, it’s time to delve into the fun, fun world of\n",
    "dealing with biological file formats!\n",
    "\n",
    "A usage example {#sec:orchids}\n",
    "---------------\n",
    "\n",
    "Before we jump right into parsers and everything else to do with\n",
    "Biopython, let’s set up an example to motivate everything we do and make\n",
    "life more interesting. After all, if there wasn’t any biology in this\n",
    "tutorial, why would you want you read it?\n",
    "\n",
    "Since I love plants, I think we’re just going to have to have a plant\n",
    "based example (sorry to all the fans of other organisms out there!).\n",
    "Having just completed a recent trip to our local greenhouse, we’ve\n",
    "suddenly developed an incredible obsession with Lady Slipper Orchids (if\n",
    "you wonder why, have a look at some [Lady Slipper Orchids photos on\n",
    "Flickr](http://www.flickr.com/search/?q=lady+slipper+orchid&s=int&z=t),\n",
    "or try a [Google Image\n",
    "Search](http://images.google.com/images?q=lady%20slipper%20orchid)).\n",
    "\n",
    "Of course, orchids are not only beautiful to look at, they are also\n",
    "extremely interesting for people studying evolution and systematics. So\n",
    "let’s suppose we’re thinking about writing a funding proposal to do a\n",
    "molecular study of Lady Slipper evolution, and would like to see what\n",
    "kind of research has already been done and how we can add to that.\n",
    "\n",
    "After a little bit of reading up we discover that the Lady Slipper\n",
    "Orchids are in the Orchidaceae family and the Cypripedioideae sub-family\n",
    "and are made up of 5 genera: *Cypripedium*, *Paphiopedilum*,\n",
    "*Phragmipedium*, *Selenipedium* and *Mexipedium*.\n",
    "\n",
    "That gives us enough to get started delving for more information. So,\n",
    "let’s look at how the Biopython tools can help us. We’ll start with\n",
    "sequence parsing in Section \\[sec:sequence-parsing\\], but the orchids\n",
    "will be back later on as well - for example we’ll search PubMed for\n",
    "papers about orchids and extract sequence data from GenBank in\n",
    "Chapter \\[chapter:entrez\\], extract data from Swiss-Prot from certain\n",
    "orchid proteins in Chapter \\[chapter:swiss\\_prot\\], and work with\n",
    "ClustalW multiple sequence alignments of orchid proteins in\n",
    "Section \\[sec:align\\_clustal\\].\n",
    "\n",
    "Parsing sequence file formats {#sec:sequence-parsing}\n",
    "-----------------------------\n",
    "\n",
    "A large part of much bioinformatics work involves dealing with the many\n",
    "types of file formats designed to hold biological data. These files are\n",
    "loaded with interesting biological data, and a special challenge is\n",
    "parsing these files into a format so that you can manipulate them with\n",
    "some kind of programming language. However the task of parsing these\n",
    "files can be frustrated by the fact that the formats can change quite\n",
    "regularly, and that formats may contain small subtleties which can break\n",
    "even the most well designed parsers.\n",
    "\n",
    "We are now going to briefly introduce the `Bio.SeqIO` module – you can\n",
    "find out more in Chapter \\[chapter:Bio.SeqIO\\]. We’ll start with an\n",
    "online search for our friends, the lady slipper orchids. To keep this\n",
    "introduction simple, we’re just using the NCBI website by hand. Let’s\n",
    "just take a look through the nucleotide databases at NCBI, using an\n",
    "Entrez online search\n",
    "(<http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?db=Nucleotide>) for\n",
    "everything mentioning the text Cypripedioideae (this is the subfamily of\n",
    "lady slipper orchids).\n",
    "\n",
    "When this tutorial was originally written, this search gave us only 94\n",
    "hits, which we saved as a FASTA formatted text file and as a GenBank\n",
    "formatted text file (files\n",
    "[ls\\_orchid.fasta](http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta)\n",
    "and\n",
    "[ls\\_orchid.gbk](http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk),\n",
    "also included with the Biopython source code under\n",
    "<span>docs/tutorial/examples/</span>).\n",
    "\n",
    "If you run the search today, you’ll get hundreds of results! When\n",
    "following the tutorial, if you want to see the same list of genes, just\n",
    "download the two files above or copy them from `docs/examples/` in the\n",
    "Biopython source code. In\n",
    "Section \\[sec:connecting-with-biological-databases\\] we will look at how\n",
    "to do a search like this from within Python.\n",
    "\n",
    "### Simple FASTA parsing example {#sec:fasta-parsing}\n",
    "\n",
    "If you open the lady slipper orchids FASTA file\n",
    "[ls\\_orchid.fasta](http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta)\n",
    "in your favourite text editor, you’ll see that the file starts like\n",
    "this:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "It contains 94 records, each has a line starting with “`>`”\n",
    "(greater-than symbol) followed by the sequence on one or more lines. Now\n",
    "try this in Python:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You should get something like this on your screen:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Notice that the FASTA format does not specify the alphabet, so\n",
    "`Bio.SeqIO` has defaulted to the rather generic `SingleLetterAlphabet()`\n",
    "rather than something DNA specific.\n",
    "\n",
    "### Simple GenBank parsing example\n",
    "\n",
    "Now let’s load the GenBank file\n",
    "[ls\\_orchid.gbk](http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk)\n",
    "instead - notice that the code to do this is almost identical to the\n",
    "snippet used above for the FASTA file - the only difference is we change\n",
    "the filename and the format string:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "This should give:\n",
    "\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 0
}