{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Introduction {#chapter:introduction}\n",
    "============\n",
    "\n",
    "What is Biopython?\n",
    "------------------\n",
    "\n",
    "The Biopython Project is an international association of developers of\n",
    "freely available Python (<http://www.python.org>) tools for\n",
    "computational molecular biology. Python is an object oriented,\n",
    "interpreted, flexible language that is becoming increasingly popular for\n",
    "scientific computing. Python is easy to learn, has a very clear syntax\n",
    "and can easily be extended with modules written in C, C++ or FORTRAN.\n",
    "\n",
    "The Biopython web site (<http://www.biopython.org>) provides an online\n",
    "resource for modules, scripts, and web links for developers of\n",
    "Python-based software for bioinformatics use and research. Basically,\n",
    "the goal of Biopython is to make it as easy as possible to use Python\n",
    "for bioinformatics by creating high-quality, reusable modules and\n",
    "classes. Biopython features include parsers for various Bioinformatics\n",
    "file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online\n",
    "services (NCBI, Expasy,...), interfaces to common and not-so-common\n",
    "programs (Clustalw, DSSP, MSMS...), a standard sequence class, various\n",
    "clustering modules, a KD tree data structure etc. and even\n",
    "documentation.\n",
    "\n",
    "Basically, we just like to program in Python and want to make it as easy\n",
    "as possible to use Python for bioinformatics by creating high-quality,\n",
    "reusable modules and scripts.\n",
    "\n",
    "What can I find in the Biopython package\n",
    "----------------------------------------\n",
    "\n",
    "The main Biopython releases have lots of functionality, including:\n",
    "\n",
    "-   The ability to parse bioinformatics files into Python utilizable\n",
    "    data structures, including support for the following formats:\n",
    "\n",
    "    -   Blast output – both from standalone and WWW Blast\n",
    "\n",
    "    -   Clustalw\n",
    "\n",
    "    -   FASTA\n",
    "\n",
    "    -   GenBank\n",
    "\n",
    "    -   PubMed and Medline\n",
    "\n",
    "    -   ExPASy files, like Enzyme and Prosite\n",
    "\n",
    "    -   SCOP, including ‘dom’ and ‘lin’ files\n",
    "\n",
    "    -   UniGene\n",
    "\n",
    "    -   SwissProt\n",
    "\n",
    "-   Files in the supported formats can be iterated over record by record\n",
    "    or indexed and accessed via a Dictionary interface.\n",
    "\n",
    "-   Code to deal with popular on-line bioinformatics destinations such\n",
    "    as:\n",
    "\n",
    "    -   NCBI – Blast, Entrez and PubMed services\n",
    "\n",
    "    -   ExPASy – Swiss-Prot and Prosite entries, as well as Prosite\n",
    "        searches\n",
    "\n",
    "-   Interfaces to common bioinformatics programs such as:\n",
    "\n",
    "    -   Standalone Blast from NCBI\n",
    "\n",
    "    -   Clustalw alignment program\n",
    "\n",
    "    -   EMBOSS command line tools\n",
    "\n",
    "-   A standard sequence class that deals with sequences, ids on\n",
    "    sequences, and sequence features.\n",
    "\n",
    "-   Tools for performing common operations on sequences, such as\n",
    "    translation, transcription and weight calculations.\n",
    "\n",
    "-   Code to perform classification of data using k Nearest Neighbors,\n",
    "    Naive Bayes or Support Vector Machines.\n",
    "\n",
    "-   Code for dealing with alignments, including a standard way to create\n",
    "    and deal with substitution matrices.\n",
    "\n",
    "-   Code making it easy to split up parallelizable tasks into\n",
    "    separate processes.\n",
    "\n",
    "-   GUI-based programs to do basic sequence manipulations, translations,\n",
    "    BLASTing, etc.\n",
    "\n",
    "-   Extensive documentation and help with using the modules, including\n",
    "    this file, on-line wiki documentation, the web site, and the\n",
    "    mailing list.\n",
    "\n",
    "-   Integration with BioSQL, a sequence database schema also supported\n",
    "    by the BioPerl and BioJava projects.\n",
    "\n",
    "We hope this gives you plenty of reasons to download and start using\n",
    "Biopython!\n",
    "\n",
    "Installing Biopython\n",
    "--------------------\n",
    "\n",
    "All of the installation information for Biopython was separated from\n",
    "this document to make it easier to keep updated.\n",
    "\n",
    "The short version is go to our downloads page\n",
    "(<http://biopython.org/wiki/Download>), download and install the listed\n",
    "dependencies, then download and install Biopython. Biopython runs on\n",
    "many platforms (Windows, Mac, and on the various flavors of Linux and\n",
    "Unix). For Windows we provide pre-compiled click-and-run installers,\n",
    "while for Unix and other operating systems you must install from source\n",
    "as described in the included README file. This is usually as simple as\n",
    "the standard commands:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "(You can in fact skip the build and test, and go straight to the install\n",
    "– but its better to make sure everything seems to be working.)\n",
    "\n",
    "The longer version of our installation instructions covers installation\n",
    "of Python, Biopython dependencies and Biopython itself. It is available\n",
    "in PDF (<http://biopython.org/DIST/docs/install/Installation.pdf>) and\n",
    "HTML formats\n",
    "(<http://biopython.org/DIST/docs/install/Installation.html>).\n",
    "\n",
    "Frequently Asked Questions (FAQ)\n",
    "--------------------------------\n",
    "\n",
    "1.  *How do I cite Biopython in a scientific publication?*\\\n",
    "    Please cite our application note @cock2009 [Cock *et al.*, 2009] as\n",
    "    the main Biopython reference. In addition, please cite any\n",
    "    publications from the following list if appropriate, in particular\n",
    "    as a reference for specific modules within Biopython (more\n",
    "    information can be found on our website):\n",
    "\n",
    "    -   For the official project announcement: @chapman2000 [Chapman and\n",
    "        Chang, 2000];\n",
    "\n",
    "    -   For `Bio.PDB`: @hamelryck2003a [Hamelryck and Manderick, 2003];\n",
    "\n",
    "    -   For `Bio.Cluster`: @dehoon2004 [De Hoon *et al.*, 2004];\n",
    "\n",
    "    -   For `Bio.Graphics.GenomeDiagram`: @pritchard2006 [Pritchard\n",
    "        *et al.*, 2006];\n",
    "\n",
    "    -   For `Bio.Phylo` and `Bio.Phylo.PAML`: @talevich2012 [Talevich\n",
    "        *et al.*, 2012];\n",
    "\n",
    "    -   For the FASTQ file format as supported in Biopython, BioPerl,\n",
    "        BioRuby, BioJava, and EMBOSS: @cock2010 [Cock *et al.*, 2010].\n",
    "\n",
    "2.  *How should I capitalize “Biopython”? Is “BioPython” OK?*\\\n",
    "    The correct capitalization is “Biopython”, not “BioPython” (even\n",
    "    though that would have matched BioPerl, BioJava and BioRuby).\n",
    "\n",
    "3.  *What is going wrong with my print commands?*\\\n",
    "    This tutorial now uses the Python 3 style print *function*. As of\n",
    "    Biopython 1.62, we support both Python 2 and Python 3. The most\n",
    "    obvious language difference is the print *statement* in Python 2\n",
    "    became a print *function* in Python 3.\n",
    "\n",
    "    For example, this will only work under Python 2:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print \"Hello World!\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "    If you try that on Python 3 you’ll get a `SyntaxError`. Under Python\n",
    "    3 you must write:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Hello World!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "    Surprisingly that will also work on Python 2 – but only for simple\n",
    "    examples printing one thing. In general you need to add this magic\n",
    "    line to the start of your Python scripts to use the print function\n",
    "    under Python 2.6 and 2.7:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "    If you forget to add this magic import, under Python 2 you’ll see\n",
    "    extra brackets produced by trying to use the print function when\n",
    "    Python 2 is interpreting it as a print statement and a tuple.\n",
    "\n",
    "4.  *How do I find out what version of Biopython I have installed?*\\\n",
    "    Use this:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "print(Bio.__version__)\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 0
}